src/sqlfluff/dialects/dialect_postgres.py

```diff
@@ -86,6 +87,69 @@
 )
 
 
+@postgres_dialect.segment(replace=True)
+class ArrayLiteralSegment(BaseSegment):
+    """An array literal segment."""
+
+    type = "array_literal"
+    match_grammar = Sequence(
+        "ARRAY",
+        Bracketed(
+            Delimited(Ref("LiteralGrammar"), optional=False),
+            bracket_type="square",
+        ),
+    )
+
+
+postgres_dialect.replace(
+    BaseExpressionElementGrammar=OneOf(
+        Ref("LiteralGrammar"),
+        Ref("ArrayLiteralSegment"),
+        Ref("BareFunctionSegment"),
+        Ref("FunctionSegment"),
+        Ref("IntervalExpressionSegment"),
+        Ref("ColumnReferenceSegment"),
+        Ref("ExpressionSegment"),
+    )
+)
+
+
+@postgres_dialect.segment(replace=True)
+class DatatypeIdentifierSegment(BaseSegment):
```

Comment:
Did you consider replacing `DataTypeSegment` rather than `DataTypeIdentifierSegment`? Typically, in SQLFluff, `Identifier` is closer to a pure name. Structurally, this segment looks a bit more like `DataTypeIdentifierSegment`.

Reply:
I did consider this. That was my first pass, but i second-guessed it.  I can switch back to that.

Reply:
I'm not 100% sure it'll work in terms of handling array literals, but it seems cleaner in terms of fitting within the existing ANSI and Postgres dialects. Let us know how it goes!